File: src/boil/detectors/c_undeclared_identifier.go
// Detector for C undeclared identifier errors.
package detectors

// CUndeclaredIdentifierDetector detects C undeclared identifier errors
type CUndeclaredIdentifierDetector struct {
	*BaseDetector
}

func NewCUndeclaredIdentifierDetector() (*CUndeclaredIdentifierDetector, error) {
	base, err := NewBaseDetector(
		"CUndeclaredIdentifierDetector",
		100,
		map[string]string{
			"missing_c_include":  `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:[^\n]*undeclared.*note:.*is defined in header\s+['']<(?P<suggested_include>[^>]+)>['']`,
			"missing_c_function": `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+error:\s+[''](?P<identifier>[^'']+)['']\s+undeclared\s+\(first use`,
		},
		[]DetectorExample{
			{
				Name:     "undeclared_with_header_note",
				Input:    "main.c:5:12: 'NULL' undeclared here (not in a function); note: 'NULL' is defined in header '<stddef.h>'",
				ClueType: "missing_c_include",
				Context: map[string]string{
					"file_path":         "main.c",
					"suggested_include": "stddef.h",
				},
			},
			{
				Name:     "undeclared_first_use",
				Input:    "test.c:20:10: error: 'my_var' undeclared (first use",
				ClueType: "missing_c_function",
				Context: map[string]string{
					"file_path":   "test.c",
					"line_number": "20",
					"identifier":  "my_var",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &CUndeclaredIdentifierDetector{base}, nil
}

File: src/pipeline/detectors/c_undeclared_identifier.py
"""
Detector for C compilation errors for undeclared identifiers.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class CUndeclaredIdentifierDetector(Detector):
    """
    Detect C compilation errors for undeclared identifiers with include suggestions.

    Matches patterns like:
    - error: 'NULL' undeclared here (not in a function)
    - note: 'NULL' is defined in header '<stddef.h>'; did you forget to '#include <stddef.h>'?

    Also detects undeclared identifiers without include suggestions (missing functions):
    - error: 'disableRawMode' undeclared (first use in this function)
    """

    PATTERNS = {
        # Refactored to avoid multiple lazy quantifiers - uses greedy matching
        "missing_c_include": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:[^\n]*undeclared.*note:.*is defined in header\s+['\u2018]<(?P<suggested_include>[^>]+)>['\u2019]",
        "missing_c_function": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+error:\s+['\u2018](?P<identifier>[^'\u2019]+)['\u2019]\s+undeclared\s+\(first use",
    }

    EXAMPLES = [
        (
            "test.c:5:10: error: 'NULL' undeclared\nnote: 'NULL' is defined in header '<stddef.h>'",
            {
                "clue_type": "missing_c_include",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.c",
                    "suggested_include": "stddef.h",
                },
            },
        ),
        (
            "test.c:15:3: error: 'disableRawMode' undeclared (first use in this function)",
            {
                "clue_type": "missing_c_function",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.c",
                    "line_number": "15",
                    "identifier": "disableRawMode",
                },
            },
        ),
    ]
